benches/nexmark/main.rs

```diff
@@ -2,64 +2,122 @@
+
+// TODO(absoludity): Some tools mentioned at
+// https://nnethercote.github.io/perf-book/benchmarking.html but as had been
+// said earlier, most are more suited to micro-benchmarking.  I assume that our
+// best option for comparable benchmarks will be to try to do exactly what the
+// Java implementation does: core(s) * time [see Run
+// Nexmark](https://github.com/nexmark/nexmark#run-nexmark).  Right now, just
+// grab elapsed time for each query run.  See
+// https://github.com/matklad/t-cmd/blob/master/src/main.rs Also CpuMonitor.java
+// in nexmark (binary that uses procfs to get cpu usage ever 100ms?)
+
 fn main() -> Result<()> {
     let nexmark_config = NexmarkConfig::parse();
-    let generator_config = GeneratorConfig::new(nexmark_config, 0, 0, 0, 0);
-
-    let root = Root::build(|circuit| {
-        // TODO(absoludity): CPUProfiler is not currently used in any of the
-        // other benchmarks, only commented out. Not yet sure whether it will be
-        // helpful either. Some tools mentioned at
-        // https://nnethercote.github.io/perf-book/benchmarking.html but as had
-        // been said earlier, most are more suited to micro-benchmarking.  I
-        // assume that our best option for comparable benchmarks will be to try
-        // to do exactly what the Java implementation does: core(s) * time [see
-        // Run Nexmark](https://github.com/nexmark/nexmark#run-nexmark).  Right
-        // now, just grab simple timestamp to do a duration for 100_000_000 events with
-        // a source generating 1M events/sec as per the Nexmark results.
-        let start_time = std::time::SystemTime::now();
-
-        let source =
-            NexmarkSource::<ThreadRng, isize, OrdZSet<Event, isize>>::new(generator_config);
-        let input = circuit.add_source(source);
-
-        let output = q1(input);
-        let mut events_processed: u64 = 0;
-
-        output.inspect(move |zs: &OrdZSet<_, _>| {
-            events_processed += zs.len() as u64;
-            if events_processed >= 100_000_000 {
-                println!(
-                    "{} events processed in {}ms",
-                    events_processed,
-                    std::time::SystemTime::now()
-                        .duration_since(start_time)
-                        .unwrap()
-                        .as_millis()
-                );
-                // TODO: send signal through channel to stop iterating once
-                // sufficient data for test is processed?
-            }
-        });
-    })
-    .unwrap();
+    let max_events = nexmark_config.max_events;
+    let queries_to_run = nexmark_config.query.clone();
+    let generator_config = GeneratorConfig::new(nexmark_config, 0, 0, 0);
+
+    run_queries!(
+        generator_config,
+        max_events,
+        queries_to_run,
+        (String::from("q0"), q0),
+        (String::from("q1"), q1),
+        (String::from("q2"), q2),
+        (String::from("q3"), q3),
+        (String::from("q4"), q4)
```

Comment:
You should just be able to use literals here, no need for the from calls

Reply:
I wasn't able to... if I change these to literals, it works through to requiring the `query` field of the config to be `Vec<&str>` which then needs lifetimes... Or did you mean that the literals should be coerced into strings? Let me know if I've missed something.

Reply:
Ah, in that case just move the string creation into the macro by putting a `.to_string()` call wherever the strings are used

Reply:
Right, done.